src/SDL_log.c

```diff
@@ -113,59 +119,152 @@ static int SDL_android_priority[SDL_LOG_PRIORITY_COUNT] = {
 };
 #endif // SDL_PLATFORM_ANDROID
 
+static void SDLCALL SDL_LoggingChanged(void *userdata, const char *name, const char *oldValue, const char *hint)
+{
+    SDL_ResetLogPriorities();
+}
+
 void SDL_InitLog(void)
 {
-    if (!log_function_mutex) {
-        // if this fails we'll try to continue without it.
-        log_function_mutex = SDL_CreateMutex();
+    if (SDL_AtomicGet(&SDL_log_initialized)) {
+        return;
+    }
+
+    // Do a little tap dance to make sure only one thread initializes logging
```

Comment:
I had a different way of dealing with on-demand mutex creation where it might race between two threads, from the async i/o stuff. Spinlocks are kinda risky because, on the Nintendo Switch, the scheduler might decide to hand the CPU to the spinning thread, and since it never hits a syscall, it never decides to give the thing holding the lock back the CPU! I've never had this specific problem on other platforms, but spinlocks suck in general, so it's better to avoid them if possible.

```c
static SDL_Mutex *AtomicMutex(SDL_Mutex **lock)
{
    SDL_Mutex *mutex = *lock;
    if (!mutex) {
        mutex = SDL_CreateMutex();
        if (!mutex) {
            return NULL;  // oh well.
        }

        if (!SDL_AtomicCompareAndSwapPointer((void **) lock, NULL, mutex)) {
            SDL_DestroyMutex(mutex);  // something else got there first, use their mutex.
            mutex = (SDL_Mutex *) SDL_AtomicGetPointer((void **) lock);
        }
    }

    return mutex;
}
```

Once the lock is in place, this doesn't use atomics at all, and if it needs to create the lock, it doesn't spin, it just decides whether it won the race or not. The worst case is losing threads in the race make an extra mutex and destroy it immediately, but I'm guessing that's extremely uncommon.

Just in case this is useful, feel free to steal it.

Reply:
Good point. I have to delay returning until initialization is complete, but I can do that a different way. BTW, the GPU folks added a spinlock there, it might be worth bringing this up.

Reply:
Yeah, there's a couple sprinkled around SDL's codebase, still, if I recall; in theory they should be safe for setting a variable or so, but something like SDL_CreateMutex() is probably too heavyweight (makes a syscall, etc).

I'll grep around for spinlocks and see what's up.